### ISSUE TITLE
Stale bundled molecule schema rejects valid native ansible: config, and molecule.yml isn’t schema-validated by default

### ISSUE THREAD

**Issue by jeffcpullen** [ansible/ansible-lint#5119]
##### Summary

ansible-lint ships a `molecule` JSON schema but has two coupled defects: it does **not** apply that schema to `molecule.yml` by default (the file is mapped to kind `yaml`), and the schema it ships is **stale** (still the pre-native model, missing the `ansible:` block). Either alone is a problem; together they mean molecule config is effectively unvalidated by ansible-lint out of the box, and cannot be safely turned on as shipped because the stale schema false-errors on every current native config.

Relevant files:

- [`src/ansiblelint/config.py`](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/config.py) — `DEFAULT_KINDS` maps `molecule.yml` to kind `yaml`
- [`src/ansiblelint/schemas/molecule.json`](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/schemas/molecule.json) — the stale bundled schema
- [`src/ansiblelint/schemas/__store__.json`](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/schemas/__store__.json) — self-referential store entry, the drift root cause
- [`ansible-community/molecule` `src/molecule/data/molecule.json`](https://github.com/ansible-community/molecule/blob/main/src/molecule/data/molecule.json) — the authoritative up-to-date schema

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console (paste below)
ansible-lint 26.6.0 using ansible-core:2.21.2 ansible-compat:26.6.0 ruamel-yaml:0.19.1 ruamel-yaml-clib:0.2.15
```

- ansible installation method: pip
- ansible-lint installation method: pip

Also relevant: molecule 26.6.0 (native config model, current since molecule v25.9.0).

##### STEPS TO REPRODUCE

Two files, one command.

1. `molecule/default/molecule.yml` — a valid ansible-native config, accepted by molecule 26.6.0:

```yaml
---
dependency:
  name: galaxy
ansible:
  executor:
    backend: ansible-playbook
  playbooks:
    converge: converge.yml
scenario:
  name: default
  test_sequence:
    - converge
```

2. `.ansible-lint` — classify `molecule.yml` as kind `molecule`, as ansible-lint's own error text suggests ("use `kinds:` option in linter config to help it pick correct file type"):

```yaml
---
kinds:
  - molecule: "**/molecule/*/molecule.yml"
```

3. Run (with `--offline` so the bundled schema, not a network refresh, is exercised):

```console (paste below)
ansible-lint --offline molecule/default/molecule.yml
```

Then delete the `.ansible-lint` override and run the same command again.

##### Desired Behavior

`molecule.yml` should be schema-validated against an up-to-date molecule schema by default, and a valid ansible-native molecule config should pass.

Requested change, two parts, **in this order** (order matters — part 2 before part 1 would break every native molecule user with false errors):

1. **Fix the schema.** Sync `src/ansiblelint/schemas/molecule.json` to the ansible-native model from Molecule's authoritative [`src/molecule/data/molecule.json`](https://github.com/ansible-community/molecule/blob/main/src/molecule/data/molecule.json), and ideally have `__store__.json` track Molecule's schema URL so it cannot drift again.
2. **Wire it up.** Map the molecule config file to kind `molecule` in `config.py` `DEFAULT_KINDS` so `molecule.yml` is actually validated against the now-correct schema.

If kind `molecule` was intentionally downgraded to `yaml` as a stopgap for the stale schema, this is the request to finish that work rather than leave molecule config unvalidated.

##### Actual Behavior

With the `.ansible-lint` override in place, ansible-lint fatally rejects the valid native config:

```console
$ ansible-lint --offline molecule/default/molecule.yml
schema[molecule]: $ Additional properties are not allowed ('ansible' was unexpected). See https://docs.ansible.com/projects/molecule/configuration/
This rule is not skippable and stops further processing of the file.
Failed: 1 failure(s), 0 warning(s)
$ echo $?
2
```

Without the override, the same file passes with **no schema check at all** (kind `yaml`):

```console
$ rm .ansible-lint
$ ansible-lint --offline molecule/default/molecule.yml
Passed: 0 failure(s), 0 warning(s) in 1 files processed of 1 encountered. Last profile that met the validation criteria was 'production'.
$ echo $?
0
```

Those two runs demonstrate the two coupled defects:

**Defect 1 — `molecule.yml` is not schema-validated by default.** In [`config.py`](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/config.py), `DEFAULT_KINDS` maps the molecule config file to kind `yaml`:

```python
{"yaml": "**/molecule/*/{base,molecule}.{yaml,yml}"},  # molecule config
```

No entry maps any file to kind `molecule`. Kind `yaml` has no registered JSON schema, so `molecule.yml` gets only generic YAML checks and the bundled molecule schema ([`molecule.json`](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/schemas/molecule.json), registered for kind `molecule`) is never applied out of the box. Every other structured config file gets its own specific kind (`galaxy`, `meta-runtime`, `argument_specs`, `requirements`), so molecule config is the odd one out. This is layout-agnostic — the leading `**/` matches zero or more segments, so both a standalone role's `molecule/<scenario>/molecule.yml` and a collection's `extensions/molecule/<scenario>/molecule.yml` are mapped to `yaml`; both are unvalidated.

**Defect 2 — the bundled schema is stale (pre-native).** [`molecule.json`](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/schemas/molecule.json) still has root `"required": ["driver", "platforms"]`, `"additionalProperties": false`, and no `ansible` property. Molecule has used the ansible-native model since v25.9.0 (current in 26.6.0): a top-level `ansible:` block (`cfg`/`env`/`executor`/`playbooks`) replaces `provisioner:`, and neither `driver:` nor `platforms:` is required. The `additionalProperties: false` + missing `ansible` key is what produces the fatal error above; wiring `molecule.yml` to kind `molecule` makes the stale schema reject every valid native scenario. This is a non-skippable rule.

This is the bundled copy, not a network fetch: the same fatal error reproduces with networking fully disabled (`ansible-lint --offline` under `--network none`), and the installed schema (`site-packages/ansiblelint/schemas/molecule.json`) has `additionalProperties: false`, no `ansible` key, and top-level properties `dependency, driver, lint, log, platforms, prerun, provisioner, role_name_check, scenario, verifier` — the old driver/provisioner/platforms model. Air-gapped workstations and execution environments therefore hit this stale bundled schema.

**Why they are coupled.** The schema cannot be safely turned on while it is stale (it would false-error on all current configs), which is the likely reason `molecule.yml` is left as kind `yaml`. So today the schema exists but protects nothing, and fixing either half in isolation does not fix the user experience.

**Root cause of the drift.** [`__store__.json`](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/schemas/__store__.json) maps `molecule` to ansible-lint's own raw URL (`.../ansible-lint/main/src/ansiblelint/schemas/molecule.json`), i.e. the bundled schema is self-referential and is not synced from Molecule's authoritative schema, so it has fallen behind the native transition.

**Why the bundled schema matters (connectivity).** The Red Hat Ansible VS Code extension associates `molecule.yml` with Molecule's own native schema by pointing `contributes.yamlValidation` at a remote raw-GitHub URL (`https://raw.githubusercontent.com/ansible-community/molecule/main/src/molecule/data/molecule.json`), with no bundled fallback, so an editor *with network access* validates against the current model. But that path depends on live access to GitHub. With no network (an air-gapped or disconnected workstation, or an execution environment), the language server cannot retrieve the remote schema: it reports `Unable to load schema '<url>': No content` and performs **no** schema validation, so an invalid `molecule.yml` passes unchecked and a valid one gets no validation either. `ansible-lint --offline` likewise cannot fetch a remote schema and falls back to the bundled copy, which is exactly the stale one this report is about. So the bundled schema's correctness is not redundant with the editor experience: it is the only schema-based validation available precisely where remote schema fetching is unavailable.

(Verified headlessly by driving `yaml-language-server` over LSP with the molecule schema association: online it emits correct `yaml-schema: Molecule Scenario Schema` diagnostics; under `--network none` with no cached copy it emits only the `Unable to load schema … No content` load error and no schema diagnostics. The failure is a loud non-validation, not a silent skip.)

**Comment by jeffcpullen** [comment]
### Update: findings from implementing the fix

Prototyping the fix surfaced a few things the original report did not capture. Sharing them here so triage has the full picture that matches the PR #5120 . 

**1. Validate only `molecule.yml`, not `base.yml`.** The obvious implementation of Defect 1 (flip the whole `**/molecule/*/{base,molecule}.{yaml,yml}` glob to kind `molecule`) over-reaches. Molecule's shared `base.yml` is a *config fragment* that Molecule merges into each scenario before validating; on its own it is legitimately incomplete (often empty), so validating it against the whole-file schema produces a fatal, non-skippable false error. The fix maps only `molecule.{yaml,yml}` to kind `molecule` and leaves `base.yml` as kind `yaml` (consistent with the `.config/molecule/config.*` subset, which is also left as `yaml`).

**2. This is a wider behavior change than "sync the schema" implies.** Because `molecule.yml` was kind `yaml`, it was never schema-validated, so remapping it turns validation **on** for an entire file class. On upgrade, configs that Molecule itself rejects will now newly emit `schema[molecule]` (a `VERY_HIGH`, non-skippable rule): unknown top-level keys, out-of-enum or third-party `driver.name` values, and a scenario file with no `platforms` and no `ansible:` block (e.g. a split base/scenario layout where `platforms` lives only in the merged fragment). This matches Molecule's own single-file validation, but it is worth an explicit heads-up, and the PR documents a per-project `kinds:` opt-out for unusual layouts.

**3. A separate bug in Molecule's own schema** While syncing to Molecule's authoritative schema I found that its current native schema leaves `platforms.items` as an open `{"type": "object"}` and has orphaned its `MoleculePlatformModel`, so per-platform shape (e.g. `networks`, `children`) is no longer validated. That is a Molecule-side bug, not an ansible-lint one, and I have a PR to address this **ansible-community/molecule#4661**. This is relevant here in two ways:
   - The right fix tracks Molecule's authoritative schema URL (matching the "prevent future drift" request), so once #4661 merges, ansible-lint inherits the restored per-platform validation automatically via the daily schema refresh, with no further change here.
   - It is *why* the fix intentionally does not hand-restore per-platform validation in ansible-lint's bundled copy: doing so would immediately re-introduce the very drift this issue is about.
   (As a consequence the two old `test/schemas/negative_test/molecule/platforms_*` fixtures can no longer fail against the current schema and are replaced with a native `ansible.executor.backend` enum negative.)

**4. Known limitation: empty scenario files.** A genuinely empty `molecule.yml` (distinct from the `base.yml` fragment in point 1) currently produces an unfriendly `None should not be valid ...` message under the native schema. Cleanly handling an empty/`None` document belongs in the shared schema-validation path and would affect all schema kinds, so I left it out of this focused fix; it can be a separate follow-up if maintainers want it.

**5. This matters for disconnected users.** The extension gives `molecule.yml` the `ansible` languageId (its `contributes.languages` filenamePatterns include `**/molecule/*/molecule.{yml,yaml}`), so the Ansible language server runs ansible-lint on it. There are two editor paths: as-you-type validation/completion (redhat.vscode-yaml) points `molecule.yml` at Molecule's schema by **remote URL** (not ansible-lint's copy) with no offline fallback, so an air-gapped editor that never cached it gets no molecule validation on that path; and lint diagnostics, where the language server runs `ansible-lint --offline` using ansible-lint's **bundled** schema. So for connected users this change does not alter the as-you-type path (they converge on the same schema), and for disconnected users the ansible-lint diagnostics path this PR fixes is what provides in-editor molecule validation at all (with ansible-lint installed and lint enabled, the default). The one residue offline is as-you-type completion/hover, which is remote-URL-only; restoring that would be a separate, optional vscode-ansible enhancement to vendor the schema locally, not an ansible-lint change.

**Comment by jeffcpullen** [comment]
I'm closing this issue because after more research and testing I do not think my original goal of enabling schema-validation for molecule by default is reasonable. It would need to be added to the DEFAULT_KINDS which would cause hard fails that can't be skipped. This sort of thing needs to be phased in slowly, not stop existing workflows without a warning.

The good news is that in the process of trying to make this work I did find some smaller wins that are worth tackling. I'll open those in one or more new issues. These are things like updating the local copy of the Molecule schema thats out of date, and getting it to refresh automatically rather than waiting for someone to update it. Plus, seeing if I can get tools like ansible-creator to scaffold the lint config to opt-in to the molecule validation that is currently possible.
